hazelcast/src/hazelcast/client/sql.cpp

```diff
@@ -0,0 +1,904 @@
+               version0 % version1)
+                .str());
+        }
+    }
+
+    assert(count1 == 0 || count0 > 0);
+
+    // no data members
+    if (count0 == 0) {
+        return boost::none;
+    }
+
+    size_t count;
+    member::version version;
+    if (count0 > count1 || (count0 == count1 && *version0 > *version1)) {
+        count = count0;
+        version = *version0;
+    } else {
+        count = count1;
+        version = *version1;
+    }
+
+    // otherwise return a random member from the larger group
+    static thread_local std::mt19937 generator;
+    std::uniform_int_distribution<int> distribution(0, count - 1);
+    auto random_member_index = distribution(generator);
+    for (const auto& m : members) {
+        if (!m.is_lite_member() && m.get_version() == version) {
+            random_member_index--;
+            if (random_member_index < 0) {
+                return m;
+            }
+        }
+    }
+
+    throw exception::runtime("query_utils::member_of_same_larger_version_group",
+                             "should never get here");
+}
+
+} // namespace impl
+
+sql_result::sql_result(
+  spi::ClientContext* client_context,
+  sql_service* service,
+  std::shared_ptr<connection::Connection> connection,
+  impl::query_id id,
+  int64_t update_count,
+  boost::optional<std::vector<sql_column_metadata>> columns_metadata,
+  boost::optional<sql_page> first_page,
+  boost::optional<bool> is_infinite_rows,
+  int32_t cursor_buffer_size)
+  : client_context_(client_context)
+  , service_(service)
+  , connection_(std::move(connection))
+  , query_id_(id)
+  , update_count_(update_count)
+  , first_page_(std::move(first_page))
+  , is_infinite_rows_(is_infinite_rows)
+  , iterator_requested_(false)
+  , closed_(false)
+  , cursor_buffer_size_(cursor_buffer_size)
+{
+    if (columns_metadata) {
+        row_metadata_.emplace(std::move(*columns_metadata));
+        assert(first_page_);
+        first_page_->row_metadata(row_metadata_.get_ptr());
+        first_page_->serialization_service(
+          &client_context_->get_serialization_service());
+    }
+}
+
+int64_t
+sql_result::update_count() const
+{
+    return update_count_;
+}
+
+bool
+sql_result::is_row_set() const
+{
+    return update_count() == -1;
+}
+
+boost::optional<bool>
+sql_result::is_infinite_rows() const
+{
+    return is_infinite_rows_;
+}
+
+sql_result::page_iterator_type
+sql_result::page_iterator()
+{
+    if (!first_page_) {
+        throw exception::illegal_state(
+          "sql_result::page_iterator",
+          "This result contains only update count");
+    }
+
+    if (iterator_requested_) {
+        throw exception::illegal_state("sql_result::page_iterator",
+                                       "Iterator can be requested only once");
+    }
+
+    iterator_requested_ = true;
+
+    return { this, std::move(first_page_) };
+}
+
+boost::future<void>
+sql_result::close()
```

Comment:
If after close, the user does sql_page_iterator::operator++, it will get error from the server side. 

if the user calls page_iterator() method, I added a check there.  9b048f61731cfe2debff76030cbcfc307b3da61d

Reply:
Yes user will get exception from server but I think it is better to do what Java does: Fetch requests won't go to server and fail immediately with user cancelled error.